File: logistics/logistics/report/job_details_metrics/job_details_metrics.js

frappe.query_reports["Job Details Metrics"] = {
    "filters": [
        {
            "fieldname": "job_details",
            "label": __("Job Details"),
            "fieldtype": "Link",
            "options": "Job Details",
            "reqd": 0,
            "default": "",
            "width": "80"
        },
        // {
        //     "fieldname": "agent",
        //     "label": __("Agent"),
        //     "fieldtype": "Data",
        //     "reqd": 0,
        //     "default": "",
        //     "width": "80"
        // },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "reqd": 0,
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "width": "80"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "reqd": 0,
            "default": frappe.datetime.get_today(),
            "width": "80"
        }
    ],

    "onload": function(report) {
        report.page.add_inner_button(__('Refresh Data'), function() {
            report.refresh();
        });
    },

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        if (column.fieldname == "profit_and_loss") {
            if (data.profit_and_loss < 0) {
                value = `<span style="color:red">${value}</span>`;
            } else {
                value = `<span style="color:green">${value}</span>`;
            }
        }
        if (column.fieldname == "sales_outstanding_amount") {
            if (flt(value) === 0) {
                value = `<span style="color:black">${value}</span>`;
            } else {
                value = `<span style="color:green">${value}</span>`;
            }
        }
        if (column.fieldname == "purchase_outstanding_amount") {
            if (flt(value) === 0) {
                value = `<span style="color:black">${value}</span>`;
            } else {
                value = `<span style="color:orange">${value}</span>`;
            }
        }
        return value;
    }
};

File: logistics/logistics/report/job_details_metrics/job_details_metrics.py
# Copyright (c) 2024, siva and contributors
# For license information, please see license.txt

from frappe.utils import flt
import frappe

def execute(filters=None):
    columns, data, chart = [], [], None

    columns = [
        {"label": "Job Details", "fieldname": "job_details", "fieldtype": "Link", "options": "Job Details", "width": 200},
        {"label": "Customer", "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 150},
        # {"label": "Agent", "fieldname": "agent", "fieldtype": "Data",  "width": 150},
        # {"label": "Sales Invoices", "fieldname": "sales_invoices", "fieldtype": "Data", "width": 270},
        {"label": "Total Credit", "fieldname": "total_credit", "fieldtype": "Currency", "width": 150},
        {"label": "Total Debit", "fieldname": "total_debit", "fieldtype": "Currency", "width": 150},
        {"label": "Profit & Loss", "fieldname": "profit_and_loss", "fieldtype": "Currency", "width": 150},
    ]

    job_details_filter = filters.get('job_details') if filters else None
    sales_person_filter = filters.get('sales_person') if filters else None
    from_date = filters.get('from_date') if filters else None
    to_date = filters.get('to_date') if filters else None

    user_permissions = frappe.permissions.get_user_permissions(frappe.session.user)
    user_company = user_permissions.get('Company', [None])[0]

    job_details_filters = {}
    if user_company:
        job_details_filters['company'] = user_company.doc
    if job_details_filter:
        job_details_filters['name'] = ['in', job_details_filter]
    if sales_person_filter:
        job_details_filters['sales_person'] = ['in', sales_person_filter]
    if from_date and to_date:
        job_details_filters['date'] = ['between', [from_date, to_date]]  

    job_details_list = frappe.get_all('Job Details', filters=job_details_filters, fields=['name', 'customer'])

    overall_credit_total = 0
    overall_debit_total = 0
    overall_profit_and_loss = 0

    has_data = False
    invoice_data = []
    pnl_data = []
    pnl_colors = []
    debit_data = []

    for job in job_details_list:
        total_credit = 0
        total_debit = 0
        customer = job.customer
        sales_invoices_list = []

        sales_invoices_filters = {'custom_job_number': job.name, 'docstatus': 1}

        sales_invoices = frappe.get_all('Sales Invoice', filters=sales_invoices_filters, fields=['name', 'base_total'])
        if sales_invoices:
            has_data = True
            sales_invoices_list = [inv.name for inv in sales_invoices]

        for inv in sales_invoices:
            total_credit += flt(inv.base_total)

        purchase_invoices_filters = {'custom_job_number': job.name, 'docstatus': 1}

        purchase_invoices = frappe.get_all('Purchase Invoice', filters=purchase_invoices_filters, fields=['name', 'base_grand_total'])
        if purchase_invoices:
            has_data = True
        for inv in purchase_invoices:
            total_debit += flt(inv.base_grand_total)

        journal_entries_filters = {'custom_job_details': job.name, 'docstatus': 1}

        journal_entries = frappe.get_all('Journal Entry', filters=journal_entries_filters, fields=['name'])
        if journal_entries:
            has_data = True
        for entry in journal_entries:
            entries = frappe.get_all('Journal Entry Account', filters={'parent': entry.name, 'debit': ['>', 0]}, fields=['account', 'debit'])
            for je in entries:
                total_debit += flt(je.debit)

        if has_data:
            profit_and_loss = total_credit - total_debit
            overall_profit_and_loss += profit_and_loss

            data.append({
                'job_details': job.name,
                'customer': customer,
                'sales_person': job.sales_person,
                'sales_invoices': ', '.join(sales_invoices_list),
                'total_credit': total_credit,
                'total_debit': total_debit,
                'profit_and_loss': profit_and_loss
            })

            overall_credit_total += total_credit
            overall_debit_total += total_debit

            invoice_data.append({
                'x': job.name,
                'y': total_credit
            })
            debit_data.append({
                'x': job.name,
                'y': total_debit
            })
            pnl_data.append({
                'x': job.name,
                'y': profit_and_loss
            })
            pnl_colors.append('red' if profit_and_loss < 0 else 'green')

    if not has_data:
        frappe.msgprint("No data found for the selected period")

    if has_data:
        data.append({
            'job_details': 'Overall Totals',
            'customer': '',
            'sales_person': '',
            'sales_invoices': '',
            'total_credit': overall_credit_total,
            'total_debit': overall_debit_total,
            'profit_and_loss': overall_profit_and_loss
        })

    if sales_person_filter or job_details_filter:
        chart = {
            "data": {
                "labels": [d['x'] for d in invoice_data],
                "datasets": [
                    {
                        "name": "Total Credit",
                        "values": [d['y'] for d in invoice_data]
                    },
                    {
                        "name": "Total Debit",
                        "values": [d['y'] for d in debit_data]
                    },
                    {
                        "name": "Profit & Loss",
                        "values": [d['y'] for d in pnl_data]
                    }
                ]
            },
            "type": "bar",
            "colors": [
                "#7cd6fd",  
                "#743ee2",  
                pnl_colors   
            ]
        }

    return columns, data, None, chart